### PR TITLE
Entropic time scheduler

### PR DESCRIPTION
**Pull request by btrentini**
### Description
Generates an entropic time schedule by remapping time based on the
cumulative information gain provided by a predictor function.

Inspired by the work from **Dejan Stancevic, Florian Handke, & Luca Ambrogioni. (2025).** 
Entropic Time Schedulers for Generative Diffusion Models (https://arxiv.org/abs/2504.13612)

This entropy rate is used to create an optimized, data-dependent time-stepping schedule for the generative process. 

### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
from bionemo.moco.schedules.inference_time_schedules import (
    EntropicInferenceSchedule, 
    TimeDirection #Optional
)

def predictor_fn(t: Tensor, x: Tensor) -> Tensor:
    # Ensure t is broadcastable for concatenation
    if t.ndim == 1:
        t = t.unsqueeze(-1)
    if t.shape[0] != x.shape[0]:
        t = t.expand(x.shape[0], -1)
    
    model_input = torch.cat([x, t], dim=-1)
    return model(model_input)

def x_0_sampler_fn(n_samples: int) -> Tensor:
    return #Noise Sampler

def x_1_sampler_fn(n_samples: int) -> Tensor:
    return #Target Sampler

inference_sched = EntropicInferenceSchedule(
    predictor=predictor_fn,
    x_0_sampler=x_0_sampler_fn,
    x_1_sampler=x_1_sampler_fn,
    nsteps=_FLOW_STEPS,
    n_schedule_points=10,      # More points -> more accurate schedule, but slower to generate
    batch_size=batch_size,
    direction=TimeDirection.UNIFIED,
    device=DEVICE,
)

schedule = inference_sched.generate_schedule().to(DEVICE)  
dts      = inference_sched.discretize().to(DEVICE)         

```

### Pre-submit Checklist

 - [x] I have tested these changes locally
 - [x] I have updated the documentation accordingly
 - [] I have added/updated tests as needed
 - [] All existing tests pass successfully

### CODE REVIEW

**Comment by btrentini** [comment]
@nvdreidenbach FYI

In my next iteration I intend to modify `_calculate_entropy_rate` to make it more general

File: .vscode/settings.json [hunk 1]
```diff
@@ -26,5 +26,25 @@
     "editor.rulers": [
         120
     ],
-    "autoDocstring.docstringFormat": "google-notypes"
+    "autoDocstring.docstringFormat": "google-notypes",
```
Comment (nvdreidenbach) [inline .vscode/settings.json]: if this was pushed accidentally before it should be removed

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py [hunk 2]
```diff
@@ -458,3 +457,160 @@ def generate_schedule(
...
+        """
+        # Detach and set requires_grad=True to compute gradients with respect to x
+        x = x.detach().requires_grad_(True)
+        
+        # Generate a random vector from the Rademacher distribution ({-1, 1})
+        epsilon = (torch.randint_like(x, 0, 2) * 2 - 1).to(x.dtype)
+        
+        # Trtansform input
+        model_input = torch.cat([x, t], dim=-1)
+        v = model(model_input)
```
Comment (nvdreidenbach) [inline sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Can this be made more general/ just take in a callable since in many cases model input will not be the input type of just concatentation.

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py [hunk 3]
```diff
@@ -458,3 +457,160 @@ def generate_schedule(
...
+            x (Tensor): A tensor of positions, shape [B, D].
+
+        Returns:
+            Tensor: The estimated divergence for each sample in the batch, shape [B].
+        """
+        # Detach and set requires_grad=True to compute gradients with respect to x
+        x = x.detach().requires_grad_(True)
+        
+        # Generate a random vector from the Rademacher distribution ({-1, 1})
+        epsilon = (torch.randint_like(x, 0, 2) * 2 - 1).to(x.dtype)
```
Comment (nvdreidenbach) [inline sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: add a generator if possible to be able to control the seed like in the other interpolants. This can likely be done in the init where maybe we can reuse the interpolants generator if not create a new one

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py [hunk 4]
```diff
@@ -458,3 +457,160 @@ def generate_schedule(
...
+        
+        num_interp_points = n_final_steps + 1 if not self.inclusive_end else n_final_steps
+
+        # Uniform_time is just a projection of the interpolated points
+        uniform_time = torch.linspace(0, 1, num_interp_points, device="cpu")
+        
+        import numpy as np 
+        entropic_schedule_np = np.interp(
+            uniform_time.numpy(), cumulative_entropy.cpu().numpy(), standard_time.cpu().numpy()
+        )
```
Comment (nvdreidenbach) [inline sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Can this be entropic_schedule_torch = torch.interp(uniform_time, cumulative_entropy, standard_time)

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py [hunk 5]
```diff
@@ -458,3 +457,160 @@ def generate_schedule(
...
+        self, 
+        nsteps: Optional[int] = None, 
+        device: Optional[Union[str, torch.device]] = None
+    ) -> Tensor:
+        """Generates the entropic time schedule."""
+        dev = device if device is not None else self.device
+        n_final_steps = nsteps if nsteps is not None else self.nsteps
+        
+        # Calculating entropic profile over {self.n_schedule_points}...
+        standard_time = torch.linspace(0, 1, self.n_schedule_points, device=dev)
```
Comment (nvdreidenbach) [inline sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: If you want to change the direction would this logic here also have to be flipped? Or does the final 1 - schedule work?

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py [hunk 6]
```diff
@@ -458,3 +457,160 @@ def generate_schedule(
...
+    Generates an entropic time schedule by remapping time based on the
+    cumulative information gain provided by a predictor function.
+    """
+    def __init__(
+        self,
+        predictor: Callable[[Tensor, Tensor], Tensor],
+        x_0_sampler: Callable[[int], Tensor],
+        x_1_sampler: Callable[[int], Tensor],
+        nsteps: int,
+        n_schedule_points: int = 100,
```
Comment (nvdreidenbach) [inline sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: nit: n_approx_entropy_points or soemthing just so its clear since nsteps and n schedule points sounds similar

File: .vscode/settings.json (hunk 1 above)
Comment (btrentini) [reply .vscode/settings.json]: D'oh. This is garbage from my VC Code studio configs. Will remove it!

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py (hunk 2 above)
Comment (btrentini) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Good catch. I'm gonna work with a callable

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py (hunk 3 above)
Comment (btrentini) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Not sure if I understand where the other interpolants are using the generator. I am only modifying the `inference_time_schedules.py` and none of the classes have generators?

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py (hunk 5 above)
Comment (btrentini) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Yes, it works fine since the cumulative entropy is already calcualted

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py (hunk 6 above)
Comment (btrentini) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Ok! Done

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py (hunk 4 above)
Comment (btrentini) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Humm - Thinking about this again; torch.nn.functional.interpolate doesn't do the same thing;  There's no equivalent solution so we can go with numpy;

https://discuss.pytorch.org/t/linear-interpolation-in-pytorch/66861

Alternatively, `torch.from_numpy(...)` but it's redundant 

Since it's a simple projection of the cumulative entropy over points, let's stick to numpy

Comment (nvdreidenbach) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: Good catch thanks

File: sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py (hunk 3 above)
Comment (nvdreidenbach) [reply sub-packages/bionemo-moco/src/bionemo/moco/schedules/inference_time_schedules.py]: You added it correctly. Since this schedule has a sampling the simple generator you added works well

**Comment by youhanl-nvidia** [comment]
lgtm. Nice works!

**Review by youhanl-nvidia — APPROVED** [review]
Looks good to me. Approved.
